store/sqlstore/team_store.go

```diff
@@ -372,45 +372,124 @@ func (s SqlTeamStore) GetByNames(names []string) ([]*model.Team, *model.AppError
 	return teams, nil
 }
 
+func (s SqlTeamStore) teamSearchQuery(term string, opts *model.TeamSearch, countQuery bool) sq.SelectBuilder {
+	var selectStr string
+	if countQuery {
+		selectStr = "count(*)"
+	} else {
+		selectStr = "*"
+	}
+
+	query := s.getQueryBuilder().
+		Select(selectStr).
+		From("Teams as t")
+
+	// Don't order or limit if getting count
+	if !countQuery {
+		query = query.OrderBy("t.DisplayName")
+
+		if opts.IsPaginated() {
+			query = query.Limit(uint64(*opts.PerPage)).Offset(uint64(*opts.Page * *opts.PerPage))
+		}
+	}
+
+	if len(term) > 0 {
+		term = sanitizeSearchTerm(term, "\\")
+		term = wildcardSearchTerm(term)
+
+		operatorKeyword := "ILIKE"
+		if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+			operatorKeyword = "LIKE"
+		}
+
+		query = query.Where(fmt.Sprintf("(Name %[1]s ? OR DisplayName %[1]s ?)", operatorKeyword), term, term)
+	}
+
+	var teamFilters sq.Sqlizer
+	var openInviteFilter sq.Sqlizer
+	if opts.AllowOpenInvite || opts.ExcludeAllowOpenInvite {
+		if opts.AllowOpenInvite {
+			openInviteFilter = sq.Eq{"AllowOpenInvite": true}
+		} else if opts.ExcludeAllowOpenInvite {
+			openInviteFilter = sq.Or{
+				sq.NotEq{"AllowOpenInvite": true},
+				sq.Eq{"AllowOpenInvite": nil},
+			}
+		}
+
+		teamFilters = openInviteFilter
+	}
+
+	var groupConstrainedFilter sq.Sqlizer
+	if opts.GroupConstrained || opts.ExcludeGroupConstrained {
+		if opts.GroupConstrained {
+			groupConstrainedFilter = sq.Eq{"GroupConstrained": true}
+		} else if opts.ExcludeGroupConstrained {
+			groupConstrainedFilter = sq.Or{
+				sq.NotEq{"GroupConstrained": true},
+				sq.Eq{"GroupConstrained": nil},
+			}
+		}
+
+		if teamFilters == nil {
+			teamFilters = groupConstrainedFilter
+		} else {
+			teamFilters = sq.And{teamFilters, groupConstrainedFilter}
```

Comment:
Should this be a `OR`?

Reply:
No this should be `and` - despite the fact that the front end use case for this does an `OR` (by inverting the filters submitted) since in general if a user requests an API asking for teams filtered by `AllowOpenInvite` AND `GroupConstrained` they would expect the API to return only results matching both of those filters right? If they want to get `GroupConstrained` OR `AllowOpenInvite` then they would have to send `IncludeGroupConstrained` AND `AllowOpenInvite` parameters...

I struggled with what the best way to do this would be and eventually settled on this but now im reconsidering...